src/Agent.Worker/DiagnosticLogManager.cs

```diff
@@ -0,0 +1,262 @@
+    {
+        public void UploadDiagnosticLogs(IExecutionContext executionContext, 
+                                         AgentJobRequestMessage message, 
+                                         DateTime jobStartTimeUtc)
+        {
+            executionContext.Debug("Starting diagnostic file upload.");
+
+            // Setup folders
+            // \_layout\_work\_temp\[jobname-support]
+            executionContext.Debug("Setting up diagnostic log folders.");
+            string tempDirectory = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), Constants.Path.TempDirectory);
+            ArgUtil.NotNullOrEmpty(tempDirectory, nameof(tempDirectory));
+            ArgUtil.Directory(tempDirectory, nameof(tempDirectory));
+
+            string supportRootFolder = Path.Combine(tempDirectory, message.JobName + "-support"); // TODO: Is JobName safe to use as a path? We could just generate a GUID as the name of our scoped folder?
+            Directory.CreateDirectory(supportRootFolder);
+
+            // \_layout\_work\_temp\[jobname-support]\files
+            executionContext.Debug("Creating diagnostic log files folder.");
+            string supportFilesFolder = Path.Combine(supportRootFolder, "files");
+            Directory.CreateDirectory(supportFilesFolder);
+           
+            // Create the environment file
+            // \_layout\_work\_temp\[jobname-support]\files\environment.txt
+            var configurationStore = HostContext.GetService<IConfigurationStore>();
+            AgentSettings settings = configurationStore.GetSettings();
+            int agentId = settings.AgentId;
+            string agentName = settings.AgentName;
+            int poolId = settings.PoolId;
+
+            executionContext.Debug("Creating diagnostic log environment file.");
+            string environmentFile = Path.Combine(supportFilesFolder, "environment.txt");
+            string content = GetEnvironmentContent(agentId, agentName, message.Tasks);
+            using (StreamWriter writer = File.CreateText(environmentFile)) 
+            {
+                writer.Write(content);
+            }
+
+            // Copy worker diag log files
+            List<string> workerDiagLogFiles = GetWorkerDiagLogFiles(HostContext.GetDirectory(WellKnownDirectory.Diag), jobStartTimeUtc);
+            executionContext.Debug($"Copying {workerDiagLogFiles.Count()} worker diag logs.");
+
+            foreach(string workerLogFile in workerDiagLogFiles)
+            {
+                ArgUtil.Directory(workerLogFile, nameof(workerLogFile));
+
+                string destination = Path.Combine(supportFilesFolder, Path.GetFileName(workerLogFile));
+                File.Copy(workerLogFile, destination);
+            }
+            
+            executionContext.Debug("Zipping diagnostic files.");
+
+            string buildNumber;
+            if (!message.Environment.Variables.TryGetValue(Constants.Variables.Build.Number, out buildNumber))
```

Comment:
we prefer use ExecutionContext.Variables :)

Reply:
Fixed :D